KafkaProducer: Backpressure for Ack'd messages

felixschlegel opened the pull request:
### Motivation

Our current implementation of polling `librdkafka` for new message acknowledgements in `KafkaProducer` did not support backpressure. Also, it made use of some `weak` references which we generally want to avoid due to performance reasons.

### Modifications

* create a new class called `KafkaBackPressurePollingSystem`
* implement backpressure for `KafkaProducer`'s poll-loop with `KafkaBackPressurePollingSystem`
* fix typo in `SwiftKafkaTests`

### TODO

- [x] write sensible tests for this backpressure mechanism and the state machine
- [x] how to handle dependency injection into our `KafkaBackPressurePollingSystem`, given that some of its dependencies are not `Sendable`
- [x] discuss if we want to expose a `Kafka.run()` method to the user or if we want to keep the poll task to ourselves (cc: @FranzBusch )

FranzBusch commented on Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift:
```diff
@@ -0,0 +1,274 @@
+    let stateMachine: StateMachine
+
+    // TODO: wrapped values are non-sendable
+    private let _client: NIOLockedValueBox<KafkaClient?>
+    private let _sequenceSource: NIOLockedValueBox<Producer.Source?>
```
Rather than having state outside of the state machine just put it into the associated values of the state. This way they are protected by the lock. If an action has to do something with those just return it in the action.

FranzBusch commented on Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift:
```diff
@@ -0,0 +1,274 @@
+            let action = self.stateMachine.nextPollLoopAction()
+
+            switch action {
+            case .poll:
```
Can we name this action `poolAndSleep`

FranzBusch commented on Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift:
```diff
@@ -0,0 +1,274 @@
+                self.client?.withKafkaHandlePointer { handle in
+                    rd_kafka_poll(handle, 0)
+                }
+                try? await Task.sleep(nanoseconds: pollIntervalNanos)
```
We shouldn't silence the error here. If our task got cancelled we should exit out of the run loop. Importantly though we probably want to finish the source then.

FranzBusch commented on Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift:
```diff
@@ -0,0 +1,274 @@
+            /// of `produceMore()` to continue producing messages.
+            case loopSuspended(CheckedContinuation<Void, Never>)
+            /// The system is shut down.
+            case shutDown
```
Can we rename this to `finished` to follow our other state machine naming

FranzBusch commented on Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift:
```diff
@@ -0,0 +1,274 @@
+    /// The state machine used by the ``KafkaBackPressurePollingSystem``.
+    struct StateMachine {
+        /// The possible actions for the poll loop.
+        enum PollLoopAction {
```
Can we move the actions right above the method that returns them. This makes it easier to understand what is going on.

FranzBusch commented on Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift:
```diff
@@ -0,0 +1,274 @@
+        /// Triggers an event in the state machine, causing a state transition.
+        ///
+        /// - Parameter event: The event to be triggered.
+        func eventTriggered(_ event: Event) {
```
Event triggered is way too broad. Please introduce separate methods for all the various events you model here e.g. `pollSuspended`, `pollCancelled` etc.

FranzBusch commented on Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift:
```diff
@@ -0,0 +1,274 @@
+import NIOConcurrencyHelpers
+import NIOCore
+
+// TODO: write test for evaluating right state transitions
```
To make this testable instead of passing in a `KafkaClient` you can pass in a closure that polls. In real code you can pass in a closure that does the actual polling and in the test you can pass something in which you can assert on

felixschlegel commented on Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift:
```diff
@@ -0,0 +1,274 @@
+    ///
+    /// - Parameter pollIntervalNanos: The poll interval in nanoseconds.
+    /// - Returns: An awaitable task representing the execution of the poll loop.
+    func run(pollIntervalNanos: UInt64) async {
```
Yeah works, will have to update our minimum Package version to macOS `.v13` / iOS `.v16` though 👍

felixschlegel commented:
Thanks for your review! I have implemented your changes and added 3 test cases 😄 
I still get warnings about conforming to `Sendable`. Please see my comments inline!

felixschlegel commented on Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift:
```diff
@@ -0,0 +1,263 @@
+    /// The state machine used by the ``KafkaBackPressurePollingSystem``.
+    struct StateMachine: Sendable {
+        /// The ``NIOAsyncSequenceProducer.Source`` used for yielding the messages to the ``NIOAsyncSequenceProducer``.
+        var sequenceSource: Producer.Source? // TODO: make sendable?
```
[`NIOAsyncSequenceProducer.Source`](https://github.com/apple/swift-nio/blob/6213ba7a06febe8fef60563a4a7d26a4085783cf/Sources/NIOCore/AsyncSequences/NIOAsyncSequenceProducer.swift#L211) is not `Sendable` at the moment which is giving us a warning here. Should I try making it `Sendable` in `swift-nio` or do we have a workaround for that in our case here?

FranzBusch replied:
The `StateMachine` doesn't necessarily need to be `Sendable` we are holding it in a `Lock` which makes it `Sendable` We just need to use `@unchecked Sendable` on the `KafkaBackPressurePollingSystem`

felixschlegel commented on Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift:
```diff
@@ -0,0 +1,263 @@
+    let stateMachineLock: NIOLockedValueBox<StateMachine>
+
+    /// Closure that takes care of polling `librdkafka` for new messages.
+    var pollClosure: (() -> Void)?
```
Atm this is a `var` as it captures `self` in `KafkaProducer` so it can get set _after_ our `KafkaProducer` has been initialized. However, having a `var` here makes us not conform to `Sendable`.

FranzBusch replied:
We should move this closure into the state machine. Furthermore, I don't see why this should capture the `KafkaProducer` the only thing that we need is capturing the `KafkaClient` to call `poll` on it

FranzBusch commented on Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift:
```diff
@@ -0,0 +1,263 @@
+    /// Closure that takes care of polling `librdkafka` for new messages.
+    var pollClosure: (() -> Void)?
+    /// The ``NIOAsyncSequenceProducer.Source`` used for yielding the messages to the ``NIOAsyncSequenceProducer``.
+    var sequenceSource: Producer.Source? {
```
Can't we pass this in the init or rather could we actually create the sequence in here and return it from a factory method?

FranzBusch commented on Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift:
```diff
@@ -0,0 +1,263 @@
+    /// - Parameter pollInterval: The desired time interval between two consecutive polls.
+    /// - Returns: An awaitable task representing the execution of the poll loop.
+    func run(pollInterval: Duration) async {
+        actionLoop: while true {
```
We don't need the label here instead we can just use `return` instead of `break` further down

FranzBusch commented on Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift:
```diff
@@ -0,0 +1,263 @@
+                do {
+                    try await Task.sleep(for: pollInterval)
+                } catch {
+                    self.stateMachineLock.withLockedValue { $0.shutDown() }
```
We probably have to call `source.finish()` here?

felixschlegel replied:
`source.finish()` gets invoked by `shutDown()` :+1:

FranzBusch replied:
Yes but that's inside the state machine and we want to have this outside the state machine! Since the state machine should not execute any side effects

FranzBusch commented on Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift:
```diff
@@ -0,0 +1,263 @@
+    }
+
+    /// The delivery report callback function that handles acknowledged messages.
+    private(set) lazy var deliveryReportCallback: (UnsafePointer<rd_kafka_message_t>?) -> Void = { messagePointer in
```
In the scope of this PR can we make a change so that we don't use `UnsafePointer` here but rather bridge the call-back to something well typed inside the method where we set the callback?

FranzBusch commented on Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift:
```diff
@@ -0,0 +1,263 @@
+            case .finished, .producing:
+                return
+            case .stopProducing(let continuation):
+                continuation?.resume()
```
We shouldn't resume continuations inside the state machine. This should be an action and resumed outside of the sate machine. In general, the state machine should not execute any side effects rather just do plain state transitions and return actions

FranzBusch commented on Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift:
```diff
@@ -0,0 +1,263 @@
+        var sequenceSource: Producer.Source? // TODO: make sendable?
+
+        /// The events that can be triggered by the ``KafkaBackPressurePollingSystem``.
+        enum Event {
```
Can we get rid of this now?

FranzBusch commented on Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift:
```diff
@@ -0,0 +1,263 @@
+                return
+            case .stopProducing(let continuation):
+                continuation?.resume()
+                fallthrough
```
Please avoid using `fallthrough` it makes the code harder to reason about. Rather prefer explicit state transition in every case.

FranzBusch commented on Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift:
```diff
@@ -0,0 +1,263 @@
+            switch self.state {
+            case .finished:
+                return
+            case .stopProducing(let existingContinuation) where existingContinuation != nil:
```
NIT
```suggestion
            case .stopProducing(.some):
```

FranzBusch commented on Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift:
```diff
@@ -0,0 +1,263 @@
+            case .finished:
+                return
+            case .stopProducing(let existingContinuation) where existingContinuation != nil:
+                fatalError("Created leaking continuation")
```
```suggestion
                fatalError("Internal state inconsistency. Run loop is running more than once")
```

FranzBusch commented on Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift:
```diff
@@ -0,0 +1,263 @@
+        }
+
+        /// Yields a new elements to the ``NIOAsyncSequenceProducer``.
+        fileprivate func yield(_ element: Element) -> Producer.Source.YieldResult {
```
Please yield outside of the state machine

FranzBusch reviewed (changes requested):
We are getting close! The state machine looks already quite good we just have to move some of the things outside of the state machine. I left some comments inline

felixschlegel commented:
@FranzBusch I have made the `KafkaBackPressurePollingSystem` generic in https://github.com/swift-server/swift-kafka-gsoc/pull/53/commits/624477c7a39c98e717f6be65ec712d734dbc6556, let me know what you think!

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,299 @@
+import NIOCore
+
+/// Our `AsyncSequence` implementation wrapping an `NIOAsyncSequenceProducer`.
+public struct KafkaAsyncSequence<Element, BackPressure: NIOAsyncSequenceProducerBackPressureStrategy>: AsyncSequence {
```
Let's not make the backpressure strategy generic but rather choose high/low for the user here

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,299 @@
+    public struct KafkaAsyncIterator: AsyncIteratorProtocol {
+        let wrappedIterator: NIOAsyncSequenceProducer<Element, BackPressure, PollingSystem>.AsyncIterator
+
+        public mutating func next() async -> Element? {
```
This probably should throw as well. From my understanding of the librd docs there are cases where the polling will throw an error

felixschlegel replied:
From [the documentation](https://docs.confluent.io/platform/current/clients/librdkafka/html/rdkafka_8h.html#ad50c431e3a29d14da534db49bd0682a4), `rd_kafka_poll` should not be an issue. However, [`rd_kafka_consumer_poll`](https://docs.confluent.io/platform/current/clients/librdkafka/html/rdkafka_8h.html#a65cc6cb9bd72c4084f074af0361ceddf) can result in errors. But I think it's better if we do that in the PR dedicated for `KafkaConsumer`, right?

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,299 @@
+    typealias Producer = NIOAsyncSequenceProducer<Element, BackPressure, KafkaPollingSystem>
+
+    /// The state machine that manages the system's state transitions.
+    let stateMachineLock: NIOLockedValueBox<StateMachine>
```
Can we make this private and rename it to `stateMachine`?

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,299 @@
+    let stateMachineLock: NIOLockedValueBox<StateMachine>
+
+    /// Closure that takes care of polling `librdkafka` for new messages.
+    var pollClosure: (() -> Void)? {
```
Why is this a var again? Can't we get this in the factory method?

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,299 @@
+    /// Initializes the ``KafkaBackPressurePollingSystem``.
+    /// Private initializer. The ``KafkaBackPressurePollingSystem`` is not supposed to be initialized directly.
+    /// It must rather be initialized using the ``KafkaBackPressurePollingSystem.createSystemAndSequence`` function.
+    /* USE createSystemAndSequence, TESTING only */ init() {
+        self.stateMachineLock = NIOLockedValueBox(StateMachine())
+    }
```
Why do we need this for testing and can't use the factory method?

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,299 @@
+            delegate: pollingSystem
+        )
+
+        pollingSystem.stateMachineLock.withLockedValue { stateMachine in
+            stateMachine.sequenceSource = acknowledgementsSourceAndSequence.source
+        }
```
Can we rather pass this in the init of the `KafkaPollingSystem`?

felixschlegel replied:
We cannot pass everything on `init` as we have a lot of circular dependencies, e.g. `NIOAsyncSequenceProducer` needs `KafkaPollingSystem` as its delegate, `KafkaPollingSystem` needs `NIOAsyncSequenceProducer.Source` to produce new elements. So at some point, we have to assign the references after initialization.

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,299 @@
+    /// - Returns: An awaitable task representing the execution of the poll loop.
+    func run(pollInterval: Duration) async {
+        self.stateMachineLock.withLockedValue { stateMachine in
+            guard stateMachine.running == false else {
+                fatalError("Poll loop must not be started more than once")
+            }
+            stateMachine.running = true
```
This should be done inside a method of the state machine. Let's add a method `run()` to the state machine as well that returns a `RunAction`.

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,299 @@
+                do {
+                    try await Task.sleep(for: pollInterval)
+                } catch {
+                    let command = self.stateMachineLock.withLockedValue { $0.shutDown() }
```
We have to re throw the error here

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,299 @@
+    /// The state machine used by the ``KafkaBackPressurePollingSystem``.
+    struct StateMachine {
+        /// A flag that determines if the ``run()`` method has already been invoked.
+        var running = false
+        /// Closure that takes care of polling `librdkafka` for new messages.
+        var pollClosure: (() -> Void)?
+        /// The ``NIOAsyncSequenceProducer.Source`` used for yielding the messages to the ``NIOAsyncSequenceProducer``.
+        var sequenceSource: Producer.Source?
```
Let's move all the state into the associated values of the cases

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,299 @@
+                        self.stateMachineLock.withLockedValue { $0.suspendLoop(continuation: continuation) }
+                    }
+                } onCancel: {
+                    let command = self.stateMachineLock.withLockedValue { $0.shutDown() }
+                    self.handleStateMachineCommand(command)
```
We have to resume the continuation with the `CancellationError` here

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,299 @@
+
+        /// Represents the commands that can be returned by a state machine
+        /// and shall be executed by the ``KafkaBackPressurePollingSystem``.
+        enum Command {
```
We can please stick to the `Action` naming and introduce separate actions for each method

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,299 @@
+        ///
+        /// - Parameter continuation: The continuation that will be resumed once we are allowed to produce again.
+        /// After resuming the continuation, our poll loop will start running again.
+        fileprivate mutating func suspendLoop(continuation: CheckedContinuation<Void, Never>) {
```
Why is this `fileprivate`?

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,299 @@
+        }
+
+        /// Shut down the state machine and finish producing elements.
+        mutating func shutDown() -> Command? {
```
```suggestion
        mutating func terminate() -> Command? {
```

FranzBusch reviewed (changes requested):
Getting close! Some more comments

felixschlegel commented on README.md:
```diff
@@ -27,6 +31,8 @@ for await acknowledgement in producer.acknowledgements {
+runTask.cancel()
+
 // Required
 await producer.shutdownGracefully()
```
#57

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,430 @@
+    /// Yield new elements to the underlying `NIOAsyncSequenceProducer`.
+    func yield(_ element: Element) {
+        self.stateMachine.withLockedValue { stateMachine in
+            switch stateMachine.state {
```
Why are we switching on the state here? We should switch on the state inside the state machine and return an action

FranzBusch commented on Tests/IntegrationTests/SwiftKafkaTests.swift:
```diff
@@ -70,6 +70,9 @@ final class SwiftKafkaTests: XCTestCase {
 
     func testProduceAndConsumeWithConsumerGroup() async throws {
         let producer = try await KafkaProducer(config: producerConfig, logger: .kafkaTest)
+        let _ = Task {
+            try await producer.run()
+        }
```
Let's use task groups here instead of unstructured tasks

felixschlegel commented on README.md:
```diff
@@ -16,19 +16,28 @@ let producer = try await KafkaProducer(
+    }
+
+    group.addTask {
+        let messageID = try await producer.sendAsync(
```
#59

FranzBusch commented on Tests/SwiftKafkaTests/KafkaProducerTests.swift:
```diff
@@ -54,120 +54,172 @@ final class KafkaProducerTests: XCTestCase {
+
+                // This subscribes to the acknowledgements stream and immediately terminates the stream.
+                // Required to kill the run task.
+                var iterator: KafkaAsyncSequence<KafkaProducer.Acknowledgement>.AsyncIterator? = producer
```
Since we dropping the iterator here the `didTerminate` callback on the delegate should be triggered which should lead to the `run()` task to terminate. Can we check that the `didTerminate` is triggered?

felixschlegel replied:
The solution I came up with now is quite simple: I am just invoking `self.pollingSystem.terminate()` inside of `KafkaProducer.shutDownGracefully`. By doing that we don't have to worry about if the user subscribes to our `acknowledgements` `AsyncSequence` or not.

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,459 @@
+    }
+
+    /// Kill the ``KafkaPollingSystem``'s poll loop and free its resources.
+    func terminate(_ error: Error? = nil) {
```
Can this be private?

felixschlegel replied:
Yes this would be no problem because we can also invoke `didTerminate()` from outside, but I thought that the name `terminate()` made more sense, also because `didTerminate()` is a delegate method.

(My implementation needs an `internal` method for terminating the `KafkaPollingSystem` because I invoke this from `KafkaPollingSystem.shutDownGracefully`)

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,459 @@
+    /// Determines the next action to be taken in the poll loop based on the current state.
+    ///
+    /// - Returns: The next action for the poll loop.
+    func nextPollLoopAction() -> KafkaPollingSystem.StateMachine.PollLoopAction {
```
Do we really need a separate method for this?

felixschlegel replied:
Not really, but given that our `stateMachine` is `private` this provides a safe way to access the poll loop's state for testing. There are not too many things we can assert on here in testing. Alternatively we can make `KafkaPollingSystem.yield(_:)` return the next action but this would also then just serve the purpose on being able to assert on the `KafkaPollingSystem`. In general all of this is quite hard as the `KafkaPollingSystem` does not expose that much information to the outside world.

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,459 @@
+        /// Actions to take after ``run()`` has been invoked on the ``KafkaPollingSystem/StateMachine``.
+        enum RunAction {
+            /// Do nothing, the poll loop is already terminated.
+            case alreadyClosed
+            /// The poll loop is already running. ``run()`` should not be invoked more than once!
+            case alreadyRunning
+            /// The poll loop can now be started.
+            case startLoop
```
One last nit here w.r.t. naming. Can you make sure all actions are named in way that they tell what do to e.g. `throwError` like you named `startLoop`

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,459 @@
+                    return .alreadyRunning
+                }
+                self.state = .idle(source: source, pollClosure: pollClosure, running: true)
+            case .producing(let source, let pollClosure, let running):
```
Can this happen? How can we be in `producing` before `run` was called?

felixschlegel replied:
I thought this might be possible e.g. when `produceMore()` gets called upon the `NIOAsyncSequenceProducerDelegate` before `run()` gets invoked

FranzBusch replied:
That won't happen. `produceMore()` only gets called after `yield()` returns `stopProducing`

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,459 @@
+                return .alreadyClosed
+            }
+
+            return .startLoop
```
Please add a return inside every individual case. That makes it easier to reason about and you don't have to scan until the end of the method

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,459 @@
+        /// The possible actions for the poll loop.
+        enum PollLoopAction {
+            /// Ask `librdkakfa` to receive new message acknowledgements at a given poll interval.
+            case pollAndSleep(pollClosure: (() -> Void)?)
```
This can't be optional right?

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,459 @@
+            case uninitialized
+            /// Initialized state (idle).
+            case idle(
+                source: Producer.Source?,
+                pollClosure: (() -> Void)?,
```
Why are the `source` and the `pollClosure` optional here?

felixschlegel replied:
You're right, this was from a previous implementation!

FranzBusch commented on Sources/SwiftKafka/KafkaPollingSystem.swift:
```diff
@@ -0,0 +1,459 @@
+            case .idle(let source, _, _), .producing(let source, _, _), .stopProducing(let source, _, _, _):
+                // We can also yield when in .stopProducing,
+                // the AsyncSequenceProducer will buffer for us
+                let yieldResult = source?.yield(element)
```
Please don't yield inside the state machine. Yielding is a side effect so please return that as an action.

FranzBusch commented on Sources/SwiftKafka/KafkaProducer.swift:
```diff
@@ -16,43 +16,13 @@ import Crdkafka
 /// - Note: When messages get published to a non-existent topic, a new topic is created using the ``KafkaTopicConfig``
 /// configuration object (only works if server has `auto.create.topics.enable` property set).
 public actor KafkaProducer {
+    public typealias Acknowledgement = Result<KafkaAcknowledgedMessage, KafkaAcknowledgedMessageError>
```
Why are we typealiasing this here? Let's not introduce this please but spell it out

FranzBusch commented on Sources/SwiftKafka/KafkaProducer.swift:
```diff
@@ -162,14 +129,25 @@ public actor KafkaProducer {
             }
         }
 
+        // Kill poll loop in polling system
+        self.pollingSystem.terminate()
```
I am not sure that this is enough. We should figure out what happens in a system where we don't poll anymore but still produce e.g. nobody consumes the acknowledgements sequence and it gets back pressured. Will rdkafka buffer then and at some point also back pressure or will we just OOM?

felixschlegel replied:
I believe [`rd_kafka_flush()`](https://docs.confluent.io/platform/current/clients/librdkafka/html/rdkafka_8h.html#aaff06c4372bce917c17f3c1a5d8b205d) got us covered here (we already use this).

felixschlegel commented:
Having back pressure makes no sense as `librdkafka` is queuing the messages regardless, closing, but keeping branch for reference
